src/middlewared/middlewared/plugins/auth.py

```diff
@@ -291,13 +295,34 @@ def get_token(self, token_id):
             return None
 
     @no_auth_required
-    @accepts(Str('username'), Str('password'))
+    @accepts()
+    async def two_factor_auth(self):
+        """
+        Returns true if two factor authorization is required for authorizing user's login.
+        """
+        return (await self.middleware.call('auth.twofactor.config'))['enabled']
+
+    @no_auth_required
+    @accepts(Str('username'), Str('password'), Str('otp_token', null=True, default=None))
     @pass_app
-    async def login(self, app, username, password):
-        """Authenticate session using username and password.
+    async def login(self, app, username, password, otp_token=None):
+        """
+        Authenticate session using username and password.
         Currently only root user is allowed.
+        `otp_token` must be specified if two factor authentication is enabled.
         """
         valid = await self.check_user(username, password)
+        twofactor_auth = await self.middleware.call('auth.twofactor.config')
+
+        if twofactor_auth['enabled']:
+            # We should validate that BOTH `check_user` AND `auth.twofactor.verify` are true and only then
```

Comment:
"We should run `auth.twofactor.verify` nevertheless of `check_user` result to prevent guessing passwords with a timing attack."